Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -113,141 +114,193 @@ end Ideal
 
 /-! ### Two-sided ideals in a matrix ring -/
 
+namespace RingCon
+variable {R n : Type*}
+
+section NonUnitalNonAssocSemiring
+variable [NonUnitalNonAssocSemiring R] [Fintype n]
+variable (n)
+
+/-- The ring congruence of matrices with entries related by `c`. -/
+def matricesOver (c : RingCon R) : RingCon (Matrix n n R) where
+  r M N := ∀ i j, c (M i j) (N i j)
+  iseqv.refl _ := fun _ _ => c.refl _
+  iseqv.symm h := fun _ _ => c.symm <| h _ _
+  iseqv.trans h₁ h₂ := fun _ _ => c.trans (h₁ _ _) (h₂ _ _)
+  add' h₁ h₂ := fun _ _ => c.add (h₁ _ _) (h₂ _ _)
+  mul' h₁ h₂ := fun _ _ => c.finset_sum _ fun _ _ => c.mul (h₁ _ _) (h₂ _ _)
+
+@[simp]
+theorem matricesOver_apply (c : RingCon R) (M N : Matrix n n R) :
+    c.matricesOver n M N ↔ ∀ i j, c (M i j) (N i j) :=
+  Iff.rfl
+
+theorem matricesOver_monotone : Monotone (matricesOver (R := R) n) :=
+  fun _ _ hc _ _ h _ _ => hc (h _ _)
+
+theorem matricesOver_injective [Nonempty n] : Function.Injective (matricesOver (R := R) n) :=
+  fun I J eq => RingCon.ext fun r s => by
+    have := congr_fun (DFunLike.congr_fun eq (Matrix.of fun _ _ => r)) (Matrix.of fun _ _ => s)
+    simpa using this
+
+theorem matricesOver_strictMono_of_nonempty [Nonempty n] :
+    StrictMono (matricesOver (R := R) n) :=
+  matricesOver_monotone n |>.strictMono_of_injective <| matricesOver_injective _
+
+@[simp]
+theorem matricesOver_bot : (⊥ : RingCon R).matricesOver n = ⊥ :=
+  eq_bot_iff.2 fun _ _ h => Matrix.ext h
+
+@[simp]
+theorem matricesOver_top : (⊤ : RingCon R).matricesOver n = ⊤ :=
+  eq_top_iff.2 fun _ _ _ _ _ => trivial
+
+open Matrix
+
+variable {n}
+
+/-- The congruence relation induced by `c` on `stdBasisMatrix i j`. -/
+def ofMatricesOver [DecidableEq n] (c : RingCon (Matrix n n R)) : RingCon R where
+  r x y := ∀ i j, c (stdBasisMatrix i j x) (stdBasisMatrix i j y)
```

Comment:
There are a whole bunch of equivalent ways to define this, which differ only on how they behave when `IsEmpty n`.